java-frontend/src/main/java/org/sonar/java/resolve/Types.java

```diff
@@ -194,4 +246,135 @@ public int compare(Type type, Type t1) {
+  private static Multimap<Type, Type> relevantParameterizations(List<Type> minimalErasedCandidates, List<Set<Type>> supertypes) {
+    Multimap<Type, Type> result = HashMultimap.create();
+    for (Set<Type> supertypesSet : supertypes) {
+      for (Type supertype : supertypesSet) {
+        Type erasedSupertype = supertype.erasure();
+        if (minimalErasedCandidates.contains(erasedSupertype)) {
+          result.put(erasedSupertype, supertype);
+        }
+      }
+    }
+    return result;
+  }
+
+  @VisibleForTesting
+  static Type best(List<Type> minimalCandidates) {
+    Collections.sort(minimalCandidates, new Comparator<Type>() {
+      // Sort minimal candidates by name with classes before interfaces, to guarantee always the same type is returned when approximated.
+      @Override
+      public int compare(Type type, Type t1) {
+        Symbol.TypeSymbol typeSymbol = type.symbol();
+        Symbol.TypeSymbol t1Symbol = t1.symbol();
+        if (typeSymbol.isInterface() && t1Symbol.isInterface()) {
+          return type.name().compareTo(t1.name());
+        } else if (typeSymbol.isInterface()) {
+          return 1;
+        } else if (t1Symbol.isInterface()) {
+          return -1;
+        }
+        return type.name().compareTo(t1.name());
+      }
+    });
+    // FIXME SONARJAVA-1632 should return union of types
+    return minimalCandidates.get(0);
+  }
+
+  /**
+   * Let the "candidate" parameterization of G, Candidate(G), be the most specific parameterization of the generic type G that contains all
+   * the relevant parameterizations of G: Candidate(G) = lcp(Relevant(G)), where lcp() is the least containing parameterization. 
+   * @param types
+   * @return
+   */
+  private Type leastContainingParameterization(List<Type> types) {
+    if (types.size() == 1) {
+      return types.get(0);
+    }
+    JavaType type1 = (JavaType) types.get(0);
+    JavaType type2 = (JavaType) types.get(1);
+    Type reduction = leastContainingTypeArgument(type1, type2);
+
+    List<Type> reducedList = Lists.newArrayList(reduction);
+    reducedList.addAll(types.subList(2, types.size()));
+
+    return leastContainingParameterization(reducedList);
+  }
+
+  private Type leastContainingTypeArgument(JavaType type1, JavaType type2) {
+    Preconditions.checkArgument(type1.isTagged(JavaType.PARAMETERIZED) && type2.isTagged(JavaType.PARAMETERIZED));
+
+    TypeSubstitution typeSubstitution1 = ((ParametrizedTypeJavaType) type1).typeSubstitution;
+    TypeSubstitution typeSubstitution2 = ((ParametrizedTypeJavaType) type2).typeSubstitution;
+
+    TypeSubstitution newTypeSubstitution = new TypeSubstitution();
+    for (TypeVariableJavaType typeVar : typeSubstitution1.typeVariables()) {
+      JavaType subs1 = typeSubstitution1.substitutedType(typeVar);
+      JavaType subs2 = typeSubstitution2.substitutedType(typeVar);
+
+      boolean isWildcard1 = subs1.isTagged(JavaType.WILDCARD);
+      boolean isWildcard2 = subs2.isTagged(JavaType.WILDCARD);
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) NullPointerException might be thrown as 'subs2' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)